coalesce: respect empty-map overrides for sub-keys when merging values

kosiew opened the pull request:
Closes #31291 

## What this PR does / why we need it

This PR fixes a bug in Helm's value coalescing where an explicitly empty map provided in an overrides file (for a *sub-key*) did not suppress the chart defaults. In other words, when a user specifies `someKey: {}` in a `values-override.yaml`, chart defaults under `someKey` could still be merged into the final values, which is surprising and inconsistent with the behavior of `--set someKey={}`.

This change makes the coalescer treat an *empty destination map* as an intentional override for non-top-level keys: if the destination map is empty and we are coalescing a sub-key (i.e. `prefix != ""`) and we are *not* in merge mode, we short-circuit and do not copy defaults from the source map. Top-level coalescing is unchanged (top-level `values` should still inherit chart defaults when appropriate).

Why this matters:

* Users commonly expect `key: {}` in an overrides YAML to remove chart defaults under `key`.
* Previously this only worked reliably when using `--set key={}`, causing an inconsistency between CLI and file-based overrides.
* Fixing this makes file-based overrides behave the same as `--set` and aligns with user expectations.

---

## Summary of changes

* **Code:** Added a short-circuit in `coalesceTablesFullKey` to return early when:

  * destination (`dst`) is empty, and
  * we're operating on a sub-key (`prefix != ""`), and
  * we are *not* merging (`!merge`).

  This prevents copying defaults from `src` into an intentionally-empty `dst` map.

* **Tests:** Added unit tests that assert the new behavior:

  * `TestCoalesceValuesEmptyMapOverride` — validates that an empty map in overrides removes default keys for that key, and `nil` removes top-level keys.
  * Subtest in `TestCoalesceTables` — validates `CoalesceTables(dst, src)` leaves an intentionally-empty `dst["config"]` empty when `src` contains defaults.

Relevant files changed:

* `pkg/chart/common/util/coalesce.go` (implementation)
* `pkg/chart/common/util/coalesce_test.go` (new unit tests)

Small illustrative diff (conceptual):

```diff
@@
-    if dst == nil {
-        return src
-    }
+    if dst == nil {
+        return src
+    }
+    // If dst is empty and we're operating on a sub-key (prefix != ""),
+    // bail out when not merging so that child charts don't inherit defaults
+    // for an explicitly empty table in the parent. We only do this for
+    // non-top-level coalescing (prefix != "") because top-level coalescing
+    // still needs to pull chart defaults across when dst is empty.
+    if len(dst) == 0 && prefix != "" && !merge {
+        return dst
+    }
```

---

## Special notes for your reviewer

* Behavior change: this modifies how empty maps are treated for sub-keys. It intentionally aligns file-based overrides behavior with `--set`-style overrides; please review for any places that intentionally relied on the previous semantics.
* The change is narrow and guarded by the `prefix != ""` and `!merge` checks to limit effect surface area.
* Unit tests have been added to guard the expected behavior. Please run the test suite and verify nothing else regresses.

---

## How to reproduce / test manually

1. Create a chart with `values.yaml` containing nested defaults, e.g.:

```yaml
config:
  enabled: true
  port: 8080
```

2. Create an override file `values-override.yaml` with an empty map under `config`:

```yaml
config: {}
```

3. Render the templates using the chart and the override file:

```bash
helm template ./mychart --values values-override.yaml
```

Expected result after this PR: no `config.enabled` or `config.port` defaults should be present in the rendered outputs for the override scope (the empty `config` should suppress defaults). Prior to this PR, those defaults might still be present.

Also test `toDelete: null` behaviour in overrides to ensure nil explicitly removes a key.

---

## Backwards compatibility

* This is a behavioral fix that makes Helm treat explicit empty maps in YAML overrides as an intentional removal of nested defaults for non-top-level keys. This could be considered a breaking change for charts or users that accidentally relied on the previous merging behavior. However, the change increases consistency with CLI `--set` behavior and is more consistent with user expectations, so it is implemented as a bug fix.

* Top-level merging behavior is intentionally preserved.

---

## If applicable

* [x] this PR contains unit tests
* [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
* [x] this PR has been tested for backwards compatibility

kosiew committed:
Add test for coalescing values with empty map overrides

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Add test for CoalesceTables with empty destination table

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Clarify behavior for empty destination tables in coalesceTablesFullKey function

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Refactor tests in coalesce_test.go to use require and assert for better error handling and readability

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

scottrigby reviewed (changes requested):
The results are not as they should be. This behavior with `--set key={}` is incorrect too.

In both Helm 3 (current release `3.19.0`) and Helm 4 dev (`main`), the template output for `--set key={}` is:
```yaml
key:
- ""
```

And this PR does replicate that behavior with a custom values file. Eg:
```yaml
# file: custom-values.yaml
key: {}
```

gives the same output as `--set key={}` now.

However, that output is not correct. We want an empty map, but instead we get a list with one empty item.

We get the same output when the custom values file contains:
```yaml
# file: custom-values.yaml
key: [""]
```

The output should be:
```yaml
key: {}
```

instead of:
```yaml
key:
- ""
```

kosiew committed:
Refactor coalesce functions to handle empty maps

Reordered coalesceValues and coalesceTablesFullKey to convert
explicit empty overrides into empty maps before merging. This
prevents default map entries from resurfacing, even for nested
tables. Additionally, the regression suite is expanded to cover
nested empty overrides, ensuring empty maps are preserved across
different override shapes.

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Update TestParseSet to handle nested empty lists in input

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Fix isExplicitEmptyOverride for empty maps handling

Enhanced the isExplicitEmptyOverride function to correctly identify
empty maps and placeholders created by --set key={}, preventing
unintentional reintroduction of chart defaults. Added regression
tests to ensure proper treatment of empty map placeholders and
validated that CLI values maintain empty maps after coalescing.

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Add helper function calls in TestMergeValuesExplicitEmptyOverrides for clarity

Signed-off-by: Siew Kam Onn <kosiew@gmail.com>

kosiew committed:
Merge branch 'main' into overwrite-31291
